File: TwisterPlots/twister.R
###############################################################################
# Twister Plots
#
# Alex Breskin (2021/5/14), Paul Zivich (2021/6/28)
##############################################################################

library(ggplot2)
library(rlang)

#### Step Ribbon ####
StatStepribbon <- ggproto("StatStepribbon",
                          Stat,
                          compute_group=function(., data, scales, direction = "hv",
                                                 yvars = c( "ymin", "ymax" ), ...)
                          {
                                  direction <- match.arg( direction, c( "hv", "vh" ) )
                                  data <- as.data.frame( data )[ order( data$x ), ]
                                  n <- nrow( data )
                                  
                                  if ( direction == "vh" ) {
                                          xs <- rep( 1:n, each = 2 )[ -2 * n ]
                                          ys <- c( 1, rep( 2:n, each = 2 ) )
                                  } else {
                                          ys <- rep( 1:n, each = 2 )[ -2 * n ]
                                          xs <- c( 1, rep( 2:n, each = 2))
                                  }
                                  
                                  data.frame(
                                          x = data$x[ xs ]
                                          , data[ ys, yvars, drop=FALSE ]
                                          , data[ xs, setdiff( names( data ), c( "x", yvars ) ), drop=FALSE ]
                                  )
                          },
                          required_aes=c( "x", "ymin", "ymax" ),
                          default_geom=GeomRibbon,
                          default_aes=aes( x=..x.., ymin = ..y.., ymax=Inf )
)

stat_stepribbon = function( mapping=NULL, data=NULL, geom="ribbon",
                            position="identity") {
        layer(stat=StatStepribbon, mapping=mapping, data=data, geom=geom, position=position )
}


#### Twister Plot####

#' Twister Plot
#'
#' @param dat A \code{data.frame} with the risk difference, upper and lower confidence limits, and times
#' @param xvar The variable name for the risk difference. Defaults to RD.
#' @param lcl  The variable name for the lower confidence limit of the risk difference. Defaults to RD_LCL.
#' @param ucl  The variable name for the upper confidence limit of the risk difference. Defaults to RD_UCL.
#' @param yvar The variable name for time. Defaults to "t".
#' @param xlab The x-axis label. Defaults to "Risk Difference".
#' @param ylab The y-axis label. Defaults to "Days".
#' @param treat_labs A vector containing the names of the treatment groups. Defaults to c("Treat", "Control")
#'
#' @return a \code{ggplot} object
#' @export
#'
#' @examples
#' twister_plot(dat, treat_labs = c("Vaccine", "Placebo"))
#' 
twister_plot <- function(dat,
                         xvar = RD,
                         lcl = RD_LCL,
                         ucl = RD_UCL,
                         yvar = t,
                         xlab = "Risk Difference",
                         ylab = "Days",
                         reference_line = 0.0,
                         log_scale = FALSE,
                         treat_labs = c("Vaccine", "Control")){
        
        base_breaks <- function(n = 10){
                function(x) {
                        axisTicks(log10(range(x, na.rm = TRUE)), log = TRUE, n = n)
                }
        }
        
        `%>%` <- magrittr::`%>%`
        pull <- dplyr::pull
        
        t_lim <- max(dat %>% pull({{yvar}}))
        if (log_scale) {
                x_lim <- max(abs(log(dat %>% pull({{lcl}}))), 
                              abs(log(dat %>% pull({{ucl}}))))
                y_scale = scale_y_continuous(limits = c(exp(-x_lim), exp(x_lim)), 
                                             trans="log", 
                                             breaks=base_breaks())
                text_loc = c(-x_lim/2, x_lim/2)
        } else {
                x_lim <- max(abs(dat %>% pull({{lcl}})), abs(dat %>% pull({{ucl}})))
                y_scale = scale_y_continuous(limits = c(-x_lim, x_lim))
                text_loc = c(-x_lim/2, x_lim/2)
        }
        
        p <- ggplot(data = dat, aes(x = {{yvar}}, y = {{xvar}})) + 
                geom_step() +
                geom_ribbon(
                        aes(ymin = {{lcl}}, ymax = {{ucl}}),
                        stat = "stepribbon",
                        alpha = 0.2,
                        direction = "hv"
                ) +
                geom_hline(yintercept = reference_line, linetype = "dotted") +
                y_scale + 
                scale_x_continuous(limits = c(0, t_lim), expand = c(0, 0)) +
                coord_flip(clip = "off") +
                theme(axis.line = element_line(colour = "black"),
                      panel.grid.major = element_blank(),
                      panel.grid.minor = element_blank(),
                      panel.background = element_blank(),
                      panel.border = element_rect(colour = "black", fill = NA),
                      plot.margin = unit(c(2, 1, 1, 1), "lines")) +
                geom_text(data = head(dat, 1), 
                          label = sprintf("Favors %s", treat_labs[1]), 
                          x = t_lim+ 5, y = text_loc[1]) +
                geom_text(data = head(dat, 1), 
                          label = sprintf("Favors %s", treat_labs[2]), 
                          x = t_lim+ 5, y = text_loc[2]) +
                xlab(ylab) +
                ylab(xlab)
        p
}

#### Example Twister Plot ####

# Reading in data
setwd("/file/path/to/data/")
data <- read.csv("data_twister.csv")

# Risk Difference Twister Plot
twister_plot(dat = data,                          
             xvar = RD,
             lcl = RD_LCL,
             ucl = RD_UCL,
             yvar = t,
             treat_labs = c("Vaccine", "Control"))

# Risk Ratio Twister Plot
twister_plot(dat = data,                          
             xvar = RR,
             lcl = RR_LCL,
             ucl = RR_UCL,
             yvar = t,
             xlab = "Risk Ratio",
             reference_line = 1.0,
             log_scale = T,
             treat_labs = c("Vaccine", "Control"))

File: TwisterPlots/twister.py
########################################################################################################################
# Twister Plots
#
# Paul Zivich (2021/5/19)
########################################################################################################################

# Importing required dependencies
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


##########################
# Twister Plot Function
def twister_plot(data, xvar, lcl, ucl, yvar, xlab="Risk Difference", ylab="Days", log_scale=False, reference_line=0.0,
                 treat_labs=("Treatment", "Placebo"), treat_labs_top=True, treat_labs_spacing="\t\t\t"):
    """Function to generate a twister plot from input data. Returns matplotlib axes which can have xlims and ylims
    set to the desired levels.

    Parameters
    ----------
    data : pandas DataFrame
        Pandas dataframe with the risk difference, upper and lower confidence limits, and times
    xvar : str
        The variable/column name for the risk difference.
    lcl : str
        The variable/column name for the lower confidence limit of the risk difference.
    ucl : str
        The variable name for the upper confidence limit of the risk difference.
    yvar : str
        The variable name for time.
    xlab : str, optional
        The x-axis label. Defaults to "Risk Difference".
    ylab : str, optional
        The y-axis label. Defaults to "Days".
    treat_labs : list, set, optional
        List of strings containing the names of the treatment groups. Only the first two elements are used in the
        labels. Defaults to 'Favors Treatment' and 'Favors Placebo'.
    treat_labs_top : bool, optional
        Whether to place the `treat_labs` at the top (True) or bottom (False). Defaults to True.
    treat_labs_spacing : str, optional
        Spacing to use between the treatment group names.

    Returns
    -------
    Matplotlib axes object

    Examples
    --------

    Example of plotting functionality

    >>> ax = twister_plot(data, xvar="RD", lcl="RD_LCL", ucl="RD_UCL", yvar="t")
    >>> ax.legend(loc='lower right')  # Added legend to the lower right corner of the plot
    >>> plt.tight_layout()  # Sets spacing of the border of the plot
    >>> plt.show()  # displays the generated image

    """
    max_t = data[yvar].max()  # Extract max y value for the plot

    # Initializing plot
    fig, ax = plt.subplots(figsize=(5, 7))  # fig_size is width by height
    ax.vlines(reference_line, 0, max_t,
              colors='gray',  # Sets color to gray for the reference line
              linestyles='--',  # Sets the reference line as dashed
              label=None)  # drawing dashed reference line at RD=0

    # Step function for Risk Difference
    ax.step(data[xvar],  # Risk Difference column
            data[yvar].shift(-1).ffill(),  # time column (shift is to make sure steps occur at correct t
            # label="RD",  # Sets the label in the legend
            color='k',  # Sets the color of the line (k=black)
            where='post')
    # Shaded step function for Risk Difference confidence intervals
    ax.fill_betweenx(data[yvar],  # time column (no shift needed here)
                     data[ucl],  # upper confidence limit
                     data[lcl],  # lower confidence limit
                     label="95% CI",  # Sets the label in the legend
                     color='k',  # Sets the color of the shaded region (k=black)
                     alpha=0.2,  # Sets the transparency of the shaded region
                     step='post')

    ax2 = ax.twiny()  # Duplicate the x-axis to create a separate label
    # "test \t test".expandtabs()
    ax2.set_xlabel("Favors " + treat_labs[0] + treat_labs_spacing.expandtabs() +  # Manually create some custom spacing
                   "Favors " + treat_labs[1],  # Top x-axes label for 'favors'
                   fontdict={"size": 10})
    ax2.set_xticks([])  # Removes top x-axes tick marks
    # Option to add the 'favors' label below the first x-axes label
    if not treat_labs_top:
        ax2.xaxis.set_ticks_position('bottom')
        ax2.xaxis.set_label_position('bottom')
        ax2.spines['bottom'].set_position(('outward', 36))

    ax.set_ylim([0, max_t])  # Sets the min and max of the y-axis
    ax.set_ylabel(ylab)  # Sets the y-label
    if log_scale:
        ax.set_xscale("log")
        xlimit = np.max([np.abs(np.log(data[lcl])),
                         np.abs(np.log(data[ucl]))])  # Extract the x-limits to use
        spacing = xlimit*2 / 20  # Sets a spacing factor. 20 seems to work well enough
        ax.set_xlim([np.exp(-xlimit - spacing), np.exp(xlimit + spacing)])  # Sets the min and max of the x-axis
    else:
        xlimit = np.max([np.abs(data[lcl]), np.abs(data[ucl])])  # Extract the x-limits to use
        spacing = xlimit*2 / 20  # Sets a spacing factor. 20 seems to work well enough
        ax.set_xlim([-xlimit-spacing, xlimit+spacing])  # Sets the min and max of the x-axis

    ax.set_xlabel(xlab,  # Sets the x-axis main label (bottom label)
                  fontdict={"size": 11,  # "weight": "bold"
                            })
    return ax


##########################
# Setup data
# Reading in data
data = pd.read_csv("data_twister.csv")  # .csv read in and managed using pandas
# data.info()  # checking that it read in correctly

##########################
# Example: Difference
ax = twister_plot(data, xvar="RD", lcl="RD_LCL", ucl="RD_UCL", yvar="t",
                  treat_labs=["Vaccine", "Placebo"])

# Formatting the axes and labels
ax.legend(loc='lower right')  # Added legend to the lower right corner of the plot
ax.set_yticks([i for i in range(0, 113, 7)])  # Sets the y-axes tick marks

plt.tight_layout()  # Sets spacing of the border of the plot
# plt.savefig("twister_plot_python.png", format='png', dpi=600)  # Saves the generated figure as .png
plt.show()  # displays the generated image

##########################
# Example: Ratio
ax = twister_plot(data, xvar="RR", lcl="RR_LCL", ucl="RR_UCL", yvar="t",
                  reference_line=1.0, log_scale=True, treat_labs=["Vaccine", "Placebo"])

# Formatting the axes and labels
ax.legend(loc='lower right')  # Added legend to the lower right corner of the plot
ax.set_yticks([i for i in range(0, 113, 7)])  # Sets the y-axes tick marks
ax.set_xticks([0.1, 0.25, 1, 5, 10])  # Sets the x-axes tick marks
ax.set_xticklabels(["0.10", "0.25", "1", "5", "10"])

plt.tight_layout()  # Sets spacing of the border of the plot
# plt.savefig("twister_plot_python.png", format='png', dpi=600)  # Saves the generated figure as .png
plt.show()  # displays the generated image
